src/include/miopen/median/problem_description.hpp

```diff
@@ -0,0 +1,250 @@
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+#pragma once
+
+#include <miopen/problem_description_base.hpp>
+#include <miopen/tensor.hpp>
+
+namespace miopen {
+
+struct NetworkConfig;
+
+namespace median {
+
+struct FwdProblemDescription : public ProblemDescriptionBase
+{
+    FwdProblemDescription(const TensorDescriptor& inputDesc_,
+                          const TensorDescriptor& outputDesc_,
+                          const TensorDescriptor& indicesDesc_,
+                          const uint64_t dim_,
+                          const bool keepdim_)
+        : inputDesc(inputDesc_),
+          outputDesc(outputDesc_),
+          indicesDesc(indicesDesc_),
+          dim(dim_),
+          keepdim(keepdim_)
+    {
+        IsValidNumDims();
+        IsValidDim();
+        IsRightLength();
+        IsSameType();
+    }
+
+    const TensorDescriptor& GetInputDesc() const { return inputDesc; }
+    const TensorDescriptor& GetOutputDesc() const { return outputDesc; }
+    const TensorDescriptor& GetIndicesDesc() const { return indicesDesc; }
+    uint64_t GetDim() const { return dim; }
+    bool GetKeepDim() const { return keepdim; }
+
+    bool IsValidNumDims() const
+    {
+        if(inputDesc.GetNumDims() > 5)
+        {
+            MIOPEN_THROW(
+                miopenStatusBadParm,
+                "MedianForward: A tensor with dimensions greater than 5 is not supported yet.");
+        }
+
+        return true;
+    }
+
+    bool IsValidDim() const
+    {
+
+        if(dim >= inputDesc.GetNumDims())
+        {
+            MIOPEN_THROW(miopenStatusBadParm,
+                         "MedianForward: Invalid dim " + std::to_string(dim) +
+                             " for input tensor with " + std::to_string(inputDesc.GetNumDims()) +
+                             " dimensions");
+        }
+
+        return true;
+    }
```

Comment:
Let's support negative `dim` value. Check [this](https://github.com/moreh-dev/MIOpen/pull/38#discussion_r1675355397) and [this](https://github.com/moreh-dev/MIOpen/blob/27d66aa7da89212dbb299ce80c2705ef06426b87/src/include/miopen/cumulative_reduction/problem_description.hpp#L66)

Reply:
I understand your suggestion. However, for this particular `median` operation, `modnn` handles the `dim` parameter in a specific way: it uses `-1` to indicate an empty dimension, meaning the reduction should be performed across the entire tensor rather than a specific dimension. I'm not sure how `MIOpen` plans to handle empty dimensions, so I’ve limited support to positive dimensions while reserving negative values (specifically `-1`) to signify other purposes.

Adding support for negative dimensions would require only a simple logic change. However, unless supporting negative dimensions is explicitly required or officially decided upon, I suggest we limit support to positive dimensions and allow higher-level layers to handle negative values of param `dim`.

Reply:
I think Pytorch specification should be followed, because we and AMD both know that, like a common language. But modnn, they doesn't have any clue about it.

Reply:
I understand your intention. 

However, regarding the reference you provided—specifically the `cumsum` operation—the dim parameter is required, making it reasonable to support negative dimensions in the backend layer. For operations where `dim` is optional, how would you handle cases where it’s left unspecified? There are, of course, multiple approaches. In my opinion, the way `modnn` handles this—defaulting to `-1` and supporting only positive `dim`—is quite reasonable. 

Therefore, unless the MIOpen convention explicitly requires handling negative dimensions in the backend layer, I suggest making it an optional choice, and I suggest let's keep this code section as it is for now (because many previous MIOpen ops already use unsigned types for param `dim` [[ref]](https://github.com/ROCm/MIOpen/blob/31be57bb7c2135a845ab6201dde54ec28c0b5c28/driver/layernorm_driver.hpp#L200-L201).

Reply:
i'll add support for negative `dim`

Reply:
I understand your intention.

However, regarding the reference you provided—specifically the cumsum operation—the dim parameter is required, making it reasonable to support negative dimensions in the backend layer. For operations where dim is optional, how would you handle cases where it’s left unspecified? There are, of course, multiple approaches. In my opinion, the way modnn handles this—defaulting to -1 and supporting only positive dim—is quite reasonable.

Therefore, unless the MIOpen convention explicitly requires handling negative dimensions in the backend layer, I suggest making it an optional choice, and I suggest let's keep this code section as it is for now (because many previous MIOpen ops already use unsigned types for param dim [[ref]](https://github.com/ROCm/MIOpen/blob/31be57bb7c2135a845ab6201dde54ec28c0b5c28/driver/layernorm_driver.hpp#L200-L201).

Reply:
Moreover, one of the uses of the dim parameter in MIOpen looks like this:
```
auto reduce_size = inputDesc.GetLengths()[dim];
```

I understand that handling negative dim values is managed elsewhere before this point. However, wouldn't it be safer to explicitly prevent `dim` from being negative at this backend level and delegate the responsibility of handling negative `dim` values to a higher-level abstraction?

Reply:
Moreover, one of the uses of the dim parameter in MIOpen looks like this:
```
auto reduce_size = inputDesc.GetLengths()[dim];
```

As shown, `dim` is used as an index in a vector. I understand that handling negative dim values is managed elsewhere before this point. However, wouldn't it be safer to explicitly prevent dim from being negative at this backend level and delegate the responsibility of handling negative dim values to a higher-level abstraction?

Reply:
> many previous MIOpen ops already use unsigned types for param dim [[ref]](https://github.com/ROCm/MIOpen/blob/31be57bb7c2135a845ab6201dde54ec28c0b5c28/driver/layernorm_driver.hpp#L200-L201)

Please don't take example from MIOpenDriver, real users don't use it.

Reply:
Like what I wrote in this [comment](https://github.com/moreh-dev/MIOpen/pull/71#discussion_r1898942779), a time consuming problem in the future could be easily prevented from now.

I understand your point of let this problem handle at another layer, but we are making something like a blackbox to them. We cannot know how stupid the users will be, but we can handle it.

Reply:
> However, regarding the reference you provided—specifically the cumsum operation—the dim parameter is required, making it reasonable to support negative dimensions in the backend layer. For operations where dim is optional, how would you handle cases where it’s left unspecified? There are, of course, multiple approaches. In my opinion, the way modnn handles this—defaulting to -1 and supporting only positive dim—is quite reasonable.

Sorry, but I don't understand what you mean, we should talk directly.